substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/monitor/MultiThreadedMonitorSupport.java

```diff
@@ -207,8 +209,49 @@ private static void slowPathMonitorEnter(Object obj) {
         StackOverflowCheck.singleton().makeYellowZoneAvailable();
         VMOperationControl.guaranteeOkayToBlock("No Java synchronization must be performed within a VMOperation: if the object is already locked, the VM is deadlocked");
         try {
-            singleton().monitorEnter(obj, MonitorInflationCause.MONITOR_ENTER);
+            int monitorOffset = getMonitorOffset(obj);
+            if (monitorOffset == 0) {
+                // Slow path
+                singleton().monitorEnter(obj, MonitorInflationCause.MONITOR_ENTER);
+                return;
+            }
 
+            /*
+             * Optimized path takes advantage of the knowledge that, when a new monitor object is
+             * created, it is not shared with other threads, so we can set its state without CAS. It
+             * also has acquisitions == 1 by construction, so we don't need to set that too.
+             */
+            long current = JavaThreads.getCurrentThreadId();
+            JavaMonitor monitor = (JavaMonitor) BarrieredAccess.readObject(obj, monitorOffset);
+            try {
+                if (monitor == null) {
+                    long startTicks = JfrTicks.elapsedTicks();
+                    JavaMonitor newMonitor = newMonitorLock();
+                    newMonitor.setState(current);
+                    monitor = (JavaMonitor) UNSAFE.compareAndExchangeObject(obj, monitorOffset, null, newMonitor);
+                    if (monitor == null) {
+                        JavaMonitorInflateEvent.emit(obj, startTicks, MonitorInflationCause.MONITOR_ENTER);
+                        monitor = newMonitor;
+                        return;
+                    }
+                }
+
+                long c = monitor.getState();
+                if (c == 0) {
+                    if (monitor.compareAndSetState(0, current)) {
+                        return;
+                    }
+                } else if (c == current) {
+                    monitor.acquisitions++;
+                    return;
+                }
+
+                long startTicks = JfrTicks.elapsedTicks();
+                monitor.acquire(1);
+                JavaMonitorEnterEvent.emit(obj, monitor.latestJfrTid, startTicks);
+            } finally {
+                monitor.latestJfrTid = current;
```

Comment:
This breaks abstraction of JFR thread ids and is also not equivalent to what happens without JFR support, but it might not matter in that case. @christianhaeubl

Reply:
Once I've reintroduced the call to `tryLock` this code has become an inlined version of `JavaMonitor.monitorEnter`, so I've replaced it with a call